llvm/lib/Target/AMDGPU/SIInstrInfo.cpp

```diff
@@ -3947,14 +3947,32 @@ MachineInstr *SIInstrInfo::convertToThreeAddress(MachineInstr &MI,
       (ST.getConstantBusLimit(Opc) > 1 || !Src0->isReg() ||
        !RI.isSGPRReg(MBB.getParent()->getRegInfo(), Src0->getReg()))) {
     MachineInstr *DefMI;
-    const auto killDef = [&]() -> void {
+    const auto killDef = [&](SlotIndex NewIdx) -> void {
       const MachineRegisterInfo &MRI = MBB.getParent()->getRegInfo();
       // The only user is the instruction which will be killed.
       Register DefReg = DefMI->getOperand(0).getReg();
+
+      if (LIS) {
```

Comment:
Could we maybe just call `shrinkToUses` instead of this code?

Reply:
That's a much heavier API, and x86 doesn't use it here

Reply:
I found another bug, but we can't use shrinkToUses. It incorrectly observes the old instruction (which we also can't erase here, which is a hassle)

Reply:
Yeah, I have thought before that `convertToThreeAddress` should be responsible for deleting the old instruction.

Reply:
I managed to use it by hackily replacing the use register